docs: added example for autoscaling using KeptnMetric (#1173)
keptn/lifecycle-toolkit: examples/support/metrics/README.md

# AutoScaling with HPA and KeptnMetrics

This example shows how `KeptnMetrics` can be used as a reference for a `HorizontalPodAutoscaler` to decide when to
scale workloads up or down. To demonstrate this, the following steps will be covered in this example:

1. Deploy a sample application
2. Create a `KeptnMetric` to monitor the throttled CPU of all pods serving our application
3. Generate some load to put the application under stress
4. Deploy a `HorizontalPodAutoscaler` to scale up/down to meet our goal for the CPU-throttling metric

## Deploying the Application

To deploy the application, you can run the following command:

> ````shell
> make 1-deploy-app
> ````

Reviewer: ```suggestion
```shell
make 1-deploy-app
```
```

> This will create a namespace called `podtato-metrics`, and a deployment called `podtato-head-entry` that's

Reviewer: ```suggestion
This will create a namespace called `podtato-metrics`, and a deployment called `podtato-head-entry` that is
```

podtato-head-entry-795b4bf76c-bjlfw   1/1     Running   0          1m
```

## Creating the Metric

To create the metric that keeps track of the throttled CPU of our application, run the following command:

```shell
make 2-create-metric
```

This will create a `KeptnMetricsProvider` pointing to Prometheus
(`http://prometheus-k8s.monitoring.svc.cluster.local:9090`), and a `KeptnMetric` that retrieves
the value of the following query:

```shell
avg(rate(container_cpu_cfs_throttled_seconds_total{container="server", namespace="podtato-metrics"}[1m]))
```

To verify that the metric is wired up properly, we can retrieve it via `kubectl`:

```shell
$kubectl get keptnmetrics.metrics.keptn.sh -n podtato-metrics cpu-throttling
NAME             PROVIDER     QUERY                                                                                                       VALUE
cpu-throttling   prometheus   avg(rate(container_cpu_cfs_throttled_seconds_total{container="server", namespace="podtato-metrics"}[1m]))   0.01433336027598159
```

## Generating Load

Now that we have our application up and running, and can retrieve the `KeptnMetric` value,

> it is time to generate some load. To do so, we will create  `Job` that regularly

Reviewer: ```suggestion
it is time to generate some load. To do so, we will create a `Job` that regularly
```

we have specified for our metric. In our case, we want to ensure that the value of our metric stays
below `0.05`, and we are willing to scale up to 10 replicas of our demo application:

```yaml
apiVersion: autoscaling/v2
kind: HorizontalPodAutoscaler
metadata:
  name: podtato-metrics-hpa
  namespace: podtato-metrics
spec:
  scaleTargetRef:
    apiVersion: apps/v1
    kind: Deployment
    name: podtato-head-entry
  minReplicas: 1
  maxReplicas: 10
  metrics:
    - type: Object
      object:
        metric:
          name: cpu-throttling
        describedObject:
          apiVersion: metrics.keptn.sh/v1alpha2
          kind: KeptnMetric
          name: cpu-throttling
        target:
          type: Value
          value: "0.05"
```

The HPA can be deployed using the following command:

```shell
make 4-deploy-hpa
```

Once the HPA has been deployed, we should immediately see it scaling up the replica count of our application,
since the metric value was already above our target. You can also verify this by inspecting the current
state of the `podtato-metrics-hpa` autoscaler:

```shell
$ make get-hpa-status

> kubectl describe  horizontalpodautoscalers.autoscaling -n podtato-kubectl podtato-mint

Reviewer: `podtato-mint` typo? 😅
